src/httpauth/digest.c

```diff
@@ -410,3 +417,270 @@ int httpauth_digest_response_encode(const struct httpauth_digest_resp *resp,
 	mbuf_set_pos(mb, 0);
 	return err;
 }
+
+
+static void httpauth_digest_chall_req_destructor(void *arg)
+{
+	struct httpauth_digest_chall_req *req = arg;
+
+	mem_deref(req->realm);
+	mem_deref(req->domain);
+	mem_deref(req->nonce);
+	mem_deref(req->opaque);
+	mem_deref(req->algorithm);
+	mem_deref(req->qop);
+	mem_deref(req->charset);
+}
+
+
+static int generate_nonce(char **pnonce, const time_t ts,
+	const char *etag, const char *secret)
+{
+	struct mbuf *mb = NULL;
+	char *nonce = NULL;
+	uint8_t *hash = NULL;
```

Comment:
Maybe simply use `uint8_t hash[SHA256_DIGEST_LENGTH];` I see no need for heap usage here.